core/io/file_access_memory.cpp

```diff
@@ -35,128 +35,237 @@
 	ERR_FAIL_NULL_V(files, ERR_FILE_NOT_FOUND);
 
-	String name = fix_path(p_path);
-	//name = DirAccess::normalize_path(name);
+	String name = p_path.simplify_path();
 
-	HashMap<String, Vector<uint8_t>>::Iterator E = files->find(name);
-	ERR_FAIL_COND_V_MSG(!E, ERR_FILE_NOT_FOUND, vformat("Can't find file '%s'.", p_path));
+	if (p_mode_flags & WRITE) {
+		files->insert(name, PackedByteArray());
+	}
 
-	data = E->value.ptrw();
-	length = E->value.size();
+	HashMap<String, Vector<uint8_t>>::Iterator E = files->find(name);
+	ERR_FAIL_COND_V_MSG(!E, ERR_FILE_NOT_FOUND, vformat("Can't find file '%s'.", name));
+	current_file = name;
 	pos = 0;
 
 	return OK;
 }
 
 bool FileAccessMemory::is_open() const {
-	return data != nullptr;
+	return !current_file.is_empty();
 }
 
 void FileAccessMemory::seek(uint64_t p_position) {
-	ERR_FAIL_NULL(data);
 	pos = p_position;
 }
 
 void FileAccessMemory::seek_end(int64_t p_position) {
-	ERR_FAIL_NULL(data);
-	pos = length + p_position;
+	pos = get_length() + p_position;
 }
 
 uint64_t FileAccessMemory::get_position() const {
-	ERR_FAIL_NULL_V(data, 0);
+	ERR_FAIL_COND_V(current_file.is_empty(), 0);
 	return pos;
 }
 
 uint64_t FileAccessMemory::get_length() const {
-	ERR_FAIL_NULL_V(data, 0);
-	return length;
+	ERR_FAIL_COND_V(current_file.is_empty(), 0);
+	return files->get(current_file).size();
 }
 
 bool FileAccessMemory::eof_reached() const {
-	return pos >= length;
+	return pos >= get_length();
 }
 
 uint64_t FileAccessMemory::get_buffer(uint8_t *p_dst, uint64_t p_length) const {
+	ERR_FAIL_COND_V(current_file.is_empty(), -1);
 	ERR_FAIL_COND_V(!p_dst && p_length > 0, -1);
-	ERR_FAIL_NULL_V(data, -1);
 
-	uint64_t left = length - pos;
+	uint64_t left = get_length() - pos;
 	uint64_t read = MIN(p_length, left);
 
 	if (read < p_length) {
 		WARN_PRINT("Reading less data than requested");
 	}
 
-	memcpy(p_dst, &data[pos], read);
+	memcpy(p_dst, files->get(current_file).ptrw() + pos, read);
 	pos += read;
 
 	return read;
 }
 
 Error FileAccessMemory::get_error() const {
-	return pos >= length ? ERR_FILE_EOF : OK;
+	return pos >= get_length() ? ERR_FILE_EOF : OK;
 }
 
 void FileAccessMemory::flush() {
-	ERR_FAIL_NULL(data);
+	ERR_FAIL_COND_MSG(current_file.is_empty(), "No opened file");
 }
 
 bool FileAccessMemory::store_buffer(const uint8_t *p_src, uint64_t p_length) {
 	ERR_FAIL_COND_V(!p_src && p_length > 0, false);
 
-	uint64_t left = length - pos;
-	uint64_t write = MIN(p_length, left);
+	PackedByteArray &p = files->get(current_file);
+	uint64_t length = get_length();
+	if (pos + p_length > length) {
+		p.resize(pos + p_length);
+	}
+	uint8_t *dst = p.ptrw() + pos;
+	memcpy(dst, p_src, p_length * sizeof(uint8_t));
 
-	memcpy(&data[pos], p_src, write);
-	pos += write;
+	pos += p_length;
+	return true;
+}
 
-	ERR_FAIL_COND_V_MSG(write < p_length, false, "Writing less data than requested.");
+Error DirAccessMemory::list_dir_begin() {
```

Comment:
Why include DirAccessMemory in file_access_memory.{cpp,h}?

Reply:
I was following the https://github.com/godotengine/godot/blob/master/core/io/file_access_pack.h example, which seems like the closest match.

It also makes it explicit they share the underlying state.